app/javascript/components/student/editor/TestRunSummary.tsx

```diff
@@ -0,0 +1,98 @@
+import React, { useReducer, useEffect } from 'react'
+import { Submission, TestRunStatus } from '../Editor'
+import { TestRunChannel } from '../../../channels/testRunChannel'
+
+export type TestRun = {
+  submissionUuid: string
+  status: TestRunStatus
+  message: string
+  tests: Test[]
+}
+
+type Test = {
+  name: string
+  status: TestStatus
+  output: string
+}
+
+enum TestStatus {
+  PASS = 'pass',
+  FAIL = 'fail',
+}
+
+function reducer(state: any, action: any) {
+  switch (action.type) {
+    case 'testRun.received':
+      return action.payload
+    case 'testRun.timeout':
+      return { ...state, status: 'timeout' }
+  }
+}
+
+function Content({ testRun }: { testRun: TestRun }) {
+  switch (testRun.status) {
+    case TestRunStatus.PASS:
+    case TestRunStatus.FAIL:
+      return (
+        <>
+          {testRun.tests.map((test: Test) => (
+            <p key={test.name}>
+              name: {test.name}, status: {test.status}, output: {test.output}
+            </p>
+          ))}
+        </>
+      )
+    case TestRunStatus.ERROR:
+    case TestRunStatus.OPS_ERROR:
+      return <p>{testRun.message}</p>
+    default:
+      return <></>
+  }
+}
+
+export function TestRunSummary({
+  submission,
+  timeout,
+}: {
+  submission: Submission
+  timeout: number
+}) {
+  const RESOLVED_TEST_STATUSES = [
+    TestRunStatus.PASS,
+    TestRunStatus.FAIL,
+    TestRunStatus.ERROR,
+    TestRunStatus.OPS_ERROR,
+  ]
+  const [testRun, dispatch] = useReducer(reducer, {
+    submissionUuid: submission.uuid,
+    status: submission.testsStatus,
+    message: '',
+    tests: [],
+  })
+  const haveTestsResolved = RESOLVED_TEST_STATUSES.includes(testRun.status)
+
+  useEffect(() => {
+    const channel = new TestRunChannel(submission, (testRun: TestRun) => {
+      dispatch({ type: 'testRun.received', payload: testRun })
+    })
+
+    return () => {
+      channel.disconnect()
+    }
```

Comment:
I think you could simplify this to:

```suggestion
    return () => channel.disconnect()
```

But feel free to ignore this.

Reply:
I want to do this, too, but our linter says otherwise.

Reply:
Ah okay, then please ignore!

Reply:
Fix the linter? :)

Reply:
The linter is not broken. What @ErikSchierboom suggests isn't the same as what's written,

```typescript
() => { channel.disconnect() }

// equals to
type fn = (): void
```

```typescript
() => channel.disconnect()}

// equals to
type fn = (): ReturnType<typeof channel['disconnect']>
```

Hooks must return the former.